src/UniqueVectors.jl

```diff
@@ -166,6 +168,21 @@ function swap!(uv::UniqueVector, to::Int, from::Int)
     return uv
 end
 
+function permute!(uv::UniqueVector, perm::AbstractVector)
+    map!(uv.lookup.vals, uv.lookup.vals) do i
+        k = findfirst(isequal(i), perm)
+        something(k, 0)
+    end
```

Comment:
I think this closure could be improved by using a `let` statement on `perm` (alternatively, we could introduce FastClosures as a dependency).  See https://docs.julialang.org/en/v1/manual/performance-tips/#man-performance-captured

Reply:
I don't think this is hitting the closure problem, it should only read from `perm` not assign to it, which is the easy case. That said I always forget how to diagnose this. Adding `let` doesn't much change what `@code_warntype` prints, nor change the time at all:
```
julia> uv = UniqueVector(randn(10)); p = sortperm(rand(10));

julia> @code_warntype permute!(uv, p)
MethodInstance for permute!(::UniqueVector{Float64}, ::Vector{Int64})
  from permute!(uv::UniqueVector, perm::AbstractVector) in UniqueVectors at /Users/me/.julia/dev/UniqueVectors/src/UniqueVectors.jl:171
Arguments
  #self#::Core.Const(permute!)
  uv::UniqueVector{Float64}
  perm::Vector{Int64}
Locals
  #13::UniqueVectors.var"#13#14"{Vector{Int64}}
Body::UniqueVector{Float64}
1 ─ %1  = UniqueVectors.:(var"#13#14")::Core.Const(UniqueVectors.var"#13#14")
│   %2  = Core.typeof(perm)::Core.Const(Vector{Int64})
│   %3  = Core.apply_type(%1, %2)::Core.Const(UniqueVectors.var"#13#14"{Vector{Int64}})
│         (#13 = %new(%3, perm))
│   %5  = #13::UniqueVectors.var"#13#14"{Vector{Int64}}
│   %6  = Base.getproperty(uv, :lookup)::Dict{Float64, Int64}
│   %7  = Base.getproperty(%6, :vals)::Vector{Int64}
│   %8  = Base.getproperty(uv, :lookup)::Dict{Float64, Int64}
│   %9  = Base.getproperty(%8, :vals)::Vector{Int64}
│         UniqueVectors.map!(%5, %7, %9)
│   %11 = Base.getproperty(uv, :items)::Vector{Float64}
│         UniqueVectors.permute!(%11, perm)
└──       return uv

julia> @btime permute!($uv, $p);
  94.462 ns (1 allocation: 160 bytes)
```